File: 4 Kyu/4 kyu - Sum of Intervals.cpp
#include <vector>
#include <limits>
#include <numeric>

int sum_intervals(std::vector<std::pair<int, int>> intervals) {
    int minv = std::numeric_limits<int>::max(), maxv = 0;

    for (auto &it : intervals) {
        maxv = std::max (maxv, it.second);
        minv = std::min (minv, it.first);
    }

    std::vector<int> range (maxv - minv);

    for (auto &it : intervals)
        for (int i = it.first; i < it.second; i++)
            range[i - minv] = true;

    return std::accumulate (range.begin(), range.end(), 0);
}

File: 4 Kyu/4 kyu - Sum of Intervals.rs
fn sum_intervals(intervals: &[(i32, i32)]) -> i32 {

    let mut intervals = intervals.iter().collect::<Vec<_>>();
    intervals.sort();

    let mut sum = 0;
    let mut maxv = intervals[0].0;

    for curr in intervals {
        if curr.1 > maxv {
            sum += curr.1 - std::cmp::max (curr.0, maxv);
            maxv = curr.1;
        }
    }

    sum
}
